src/engine_base.h

```diff
@@ -64,12 +64,8 @@ struct Engine : EnginePool::PoolItem<&_engine_pool> {
 	EngineID display_last_variant = EngineID::Invalid(); ///< NOSAVE client-side-only last variant selected.
 	EngineInfo info{};
 
-	union {
-		RailVehicleInfo rail;
-		RoadVehicleInfo road;
-		ShipVehicleInfo ship;
-		AircraftVehicleInfo air;
-	} u{};
+	/* Vehicle-type specific information, not used for Disaster or Effect vehicles. */
+	std::variant<std::monostate, RailVehicleInfo, RoadVehicleInfo, ShipVehicleInfo, AircraftVehicleInfo> u{};
```

Comment:
could this be made private? (and renamed)

maybe consider its position in the struct too, for memory purposes